src/WebJobs.Script.WebHost/BlobStorageScaleMetricsRepository.cs

```diff
@@ -0,0 +1,181 @@
+
+                foreach (var monitor in monitors)
+                {
+                    List<ScaleMetrics> monitorMetrics = new List<ScaleMetrics>();
+                    JArray rawMonitorMetrics = (JArray)allMetrics[monitor.Id];
+
+                    var monitorInterfaceType = monitor.GetType().GetInterfaces().SingleOrDefault(p => p.IsGenericType && p.GetGenericTypeDefinition() == typeof(IScaleMonitor<>));
+                    if (monitorInterfaceType == null)
+                    {
+                        // we require the monitor to implement the generic interface in order to know
+                        // what type to deserialize into
+                        continue;
+                    }
+                    Type metricsType = monitorInterfaceType.GetGenericArguments()[0];
+
+                    foreach (JObject currMetrics in rawMonitorMetrics)
+                    {
+                        var deserializedMetrics = (ScaleMetrics)currMetrics.ToObject(metricsType);
+                        if ((DateTime.UtcNow - deserializedMetrics.TimeStamp) < _scaleOptions.ScaleMetricsMaxAge)
+                        {
+                            monitorMetrics.Add(deserializedMetrics);
+                        }
+                    }
+
+                    result[monitor] = monitorMetrics;
+                }
+            }
+
+            return result;
+        }
+
+        public async Task WriteAsync(IDictionary<IScaleMonitor, ScaleMetrics> monitorMetrics)
+        {
+            if (!StorageConnectionValid() || monitorMetrics.Count == 0)
+            {
+                return;
+            }
+
+            JObject allMetrics = await ReadCurrentMetricsOrDefaultAsync();
+
+            // When writing the metrics blob back after adding the current metrics, we'll be filtering out any metrics that have expired,
+            // or for functions that no longer exist, are disabled.
+            JObject newMetrics = new JObject();
+            foreach (var pair in monitorMetrics)
+            {
+                // get the current metrics for this provider if present
+                JArray currMetrics = (JArray)allMetrics[pair.Key.Id];
+                if (currMetrics == null)
+                {
+                    currMetrics = new JArray();
+                }
+
+                // Purge any out of date metrics
+                JArray filteredMetrics = new JArray();
+                foreach (var currMetric in currMetrics)
+                {
+                    var timestamp = (DateTime)currMetric[nameof(ScaleMetrics.TimeStamp)];
+                    if ((DateTime.UtcNow - timestamp) < _scaleOptions.ScaleMetricsMaxAge)
+                    {
+                        filteredMetrics.Add(currMetric);
+                    }
+                }
+                currMetrics = filteredMetrics;
+
+                if (currMetrics.Count == _scaleOptions.ScaleMetricsRetentionCount)
```

Comment:
Should this be >= and trim all those larger? Wondering if there's a scenario where this retentioncount gets reduced between host starts and now this never hits.

Reply:
In the latest iteration this code has been removed.